rse-bash.Rmd

```diff
@@ -0,0 +1,3902 @@
+```
+
+```text
+ls: cannot access 'quotes.txt': No such file or directory
+```
+
+> **Deleting Is Forever**
+>
+> The Unix shell doesn't have a trash bin that we can recover deleted
+> files from (though most graphical interfaces to Unix do).  Instead,
+> when we delete files, they are unlinked from the file system so that
+> their storage space on disk can be recycled. Tools for finding and
+> recovering deleted files do exist, but there's no guarantee they'll
+> work in any particular situation, since the computer may recycle the
+> file's disk space right away.
+
+
+If we try to remove the `thesis` directory using `rm thesis`,
+we get an error message:
+
+```shell
+$ rm thesis
+```
+
+```text
+rm: cannot remove `thesis': Is a directory
+```
+
+This happens because `rm` by default only works on files, not directories.
+
+`rm` can remove a directory *and all its contents* if we use the 
+recursive option `-r`, and it will do so *without any confirmation prompts*:
+
+```shell
+$ rm -r thesis
+```
+
+Given that there is no way to retrieve files deleted using the shell,
+`rm -r` *should be used with great caution* (you might consider adding the interactive option `rm -r -i`).
+
+### Operations with multiple files and directories
+
+Oftentimes one needs to copy or move several files at once. This can be done by providing a list of individual filenames, or specifying a naming pattern using wildcards.  
+
+> **Wildcards**
+>
+> `*` is a **wildcard**, which matches zero or more  characters.
+> Let's consider the `data-shell/molecules` directory:
+> `*.pdb` matches `ethane.pdb`, `propane.pdb`, and every
+> file that ends with '.pdb'. On the other hand, `p*.pdb` only matches
+> `pentane.pdb` and `propane.pdb`, because the 'p' at the front only
+> matches filenames that begin with the letter 'p'.
+>
+> `?` is also a wildcard, but it matches exactly one character.
+> So `?ethane.pdb` would match `methane.pdb` whereas
+> `*ethane.pdb` matches both `ethane.pdb`, and `methane.pdb`.
+>
+> Wildcards can be used in combination with each other
+> e.g. `???ane.pdb` matches three characters followed by `ane.pdb`,
+> giving `cubane.pdb  ethane.pdb  octane.pdb`.
+>
+> When the shell sees a wildcard, it expands the wildcard to create a
+> list of matching filenames *before* running the command that was
+> asked for. As an exception, if a wildcard expression does not match
+> any file, Bash will pass the expression as an argument to the command
+> as it is. For example typing `ls *.pdf` in the `molecules` directory
+> (which contains only files with names ending with `.pdb`) results in
+> an error message that there is no file called `*.pdf`.
+> However, generally commands like `wc` and `ls` see the lists of
+> file names matching these expressions, but not the wildcards
+> themselves. It is the shell, not the other programs, that deals with
+> expanding wildcards, and this is another example of orthogonal design.
+
+
+## Pipes and Filters {#bash-pipes}
```

Comment:
I stopped reviewing here - other than linking to glossary entries, everything else is copy editing, so I think we can merge this and keep moving.

Reply:
Thanks, Greg!